src/data_structures/fenwick.md

```diff
@@ -29,138 +29,110 @@ It was first described in a paper titled "A new data structure for cumulative fr
 
 The complexity of both `sum` and `increase` depend on the function $g$.
-There are many ways to choose the function $g$ such that $0 \le g(i) \le i$ for all $i$.
-For instance, the function $g(i) = i$ works, which yields $T = A$ (in which case, the summation queries are slow).
-We could also take the function $g(i) = 0$.
-This would correspond to prefix sum arrays (in which case, finding the sum of the range $[0, i]$ will only take constant time; however, updates are slow).
+There are many ways to choose the function $g$ such that $0 \le g(i) \lt i$ for all $i$.
+For instance, the function $g(i) = i-1$ works, which yields $T = A$ (in which case, the summation queries are slow).
+We could also take the function $g(i) = 1$.
+This would correspond to prefix sum arrays (in which case, finding the sum of the range $[1, i]$ will only take constant time; however, updates are slow).
 The clever part of the algorithm for Fenwick trees is how it uses a special definition of the function $g$ which can handle both operations in $O(\log N)$ time.
 
 ### Definition of $g(i)$ { data-toc-label='Definition of <script type="math/tex">g(i)</script>' }
 
-The computation of $g(i)$ is defined using the following simple operation:
-we replace all trailing $1$ bits in the binary representation of $i$ with $0$ bits.
-
-In other words, if the least significant digit of $i$ in binary is $0$, then $g(i) = i$.
-And otherwise the least significant digit is a $1$, and we take this $1$ and all other trailing $1$s and flip them.
+The computation of $g(i)$ is defined as:
+toggling of the last set $1$ bit in the binary representation of $i$.
 
 For instance we get
-
 $$\begin{align}
-g(11) = g(1011_2) = 1000_2 &= 8 \\\\
-g(12) = g(1100_2) = 1100_2 &= 12 \\\\
-g(13) = g(1101_2) = 1100_2 &= 12 \\\\
-g(14) = g(1110_2) = 1110_2 &= 14 \\\\
-g(15) = g(1111_2) = 0000_2 &= 0 \\\\
-\end{align}$$
-
-There exists a simple implementation using bitwise operations for the non-trivial operation described above:
-
-$$g(i) = i ~\&~ (i+1),$$
-
-where $\&$ is the bitwise AND operator. It is not hard to convince yourself that this solution does the same thing as the operation described above.
-
-Now, we just need to find a way to iterate over all $j$'s, such that $g(j) \le i \le j$.
-
-It is easy to see that we can find all such $j$'s by starting with $i$ and flipping the last unset bit.
-We will call this operation $h(j)$.
-For example, for $i = 10$ we have:
-
-$$\begin{align}
-10 &= 0001010_2 \\\\
-h(10) = 11 &= 0001011_2 \\\\
-h(11) = 15 &= 0001111_2 \\\\
-h(15) = 31 &= 0011111_2 \\\\
-h(31) = 63 &= 0111111_2 \\\\
-\vdots &
+g(7) = g(111_2) = 110_2 &= 6 \\\\
+g(6) = g(110_2) = 100_2 &= 4 \\\\
+g(4) = g(100_2) = 000_2 &= 0 \\\\
 \end{align}$$
 
-Unsurprisingly, there also exists a simple way to perform $h$ using bitwise operations:
+The last set bit can be extracted using $i ~\&~ (-i)$, so the operation can be expressed as:
 
-$$h(j) = j ~|~ (j+1),$$
+$$g(i) = i - (i ~\&~ (-i)).$$
```

Comment:
This can also be implemented as `r & (r-1)`, and that's what gcc will optimize it to.